File: bindings/go/causalontology/canonical.go
// Canonicalization and content-addressed identity.
//
// Implements the identity procedure of spec/identity.md:
//  1. take the object as JSON,
//  2. keep only the identity-bearing fields for its kind (with "type"
//     injected),
//  3. serialize with the JSON Canonicalization Scheme (RFC 8785),
//  4. hash with SHA-256,
//  5. identifier = scheme + ":" + lowercase hex digest.
package causalontology

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
)

// IdentityFields lists, per kind, exactly the fields that participate in
// the identity hash; everything else (id, signature, annotations) is
// excluded by construction. 2.0.0: all seventeen whole-word kinds.
var IdentityFields = map[string][]string{
	// ---- type tier ----
	"occurrent": {"label", "category", "stratum"},
	"causal_relation_object": {"causes", "effects", "mechanism", "temporal",
		"modality", "context", "refines", "skips"},
	"continuant": {"label", "category"},
	"realizable": {"kind", "bearer", "label"},
	"stratum":    {"label", "scheme", "ordinal", "unit", "governs"},
	"bridge":     {"coarse", "fine", "relation"},
	"port":       {"bearer", "label", "direction", "accepts", "realizable"},
	"conduit":    {"label", "from", "to", "carries", "transform"},
	"quality":    {"label", "datatype", "unit", "stratum"},
	// ---- token tier ----
	"token_individual":   {"instantiates", "designator", "part_of"},
	"token_occurrence":   {"instantiates", "interval", "participants", "locus", "observer"},
	"state_assertion":    {"subject", "quality", "value", "interval"},
	"token_causal_claim": {"causes", "effects", "covering_law", "actual_delay", "counterfactual"},
	// ---- provenance tier ----
	"assertion":  {"about", "source", "evidence_type", "evidence", "strength", "confidence", "timestamp", "evidenced_by"},
	"enrichment": {"about", "field", "entry", "source", "timestamp"},
	"retraction": {"retracts", "source", "timestamp"},
	"succession": {"predecessor", "successor", "timestamp"},
}

// Prefix maps each kind to its identifier scheme; whole-word (P7): the
// scheme IS the kind string.
var Prefix = func() map[string]string {
	out := map[string]string{}
	for kind := range IdentityFields {
		out[kind] = kind
	}
	return out
}()

// KindOfPrefix is the inverse of Prefix: identifier scheme to kind.
var KindOfPrefix = func() map[string]string {
	out := map[string]string{}
	for kind := range IdentityFields {
		out[kind] = kind
	}
	return out
}()

// InferKind infers an object's kind from its type field, id prefix, or
// shape, exactly as the Python binding does.
func InferKind(obj map[string]any) (string, error) {
	if raw, present := obj["type"]; present {
		if kind, ok := raw.(string); ok {
			return kind, nil
		}
	}
	if id, ok := obj["id"].(string); ok {
		if colon := strings.IndexByte(id, ':'); colon >= 0 {
			if kind, known := KindOfPrefix[id[:colon]]; known {
				return kind, nil
			}
		}
	}
	has := func(field string) bool {
		_, present := obj[field]
		return present
	}
	if has("coarse") && has("fine") {
		return "bridge", nil
	}
	if has("causes") && has("effects") {
		return "causal_relation_object", nil
	}
	if has("retracts") {
		return "retraction", nil
	}
	if has("predecessor") && has("successor") {
		return "succession", nil
	}
	if has("field") && has("entry") {
		return "enrichment", nil
	}
	if has("evidence_type") || (has("about") && has("confidence")) {
		return "assertion", nil
	}
	if has("kind") && has("bearer") {
		return "realizable", nil
	}
	return "", errors.New(
		"cannot infer kind (occurrents and continuants share a shape); pass kind explicitly")
}

// IdentityBearing returns the identity-bearing subset of an object, with
// "type" always present. An empty kind means: infer it.
func IdentityBearing(obj map[string]any, kind string) (string, map[string]any, error) {
	if kind == "" {
		inferred, err := InferKind(obj)
		if err != nil {
			return "", nil, err
		}
		kind = inferred
	}
	fields, known := IdentityFields[kind]
	if !known {
		return "", nil, fmt.Errorf("unknown kind: %q", kind)
	}
	out := map[string]any{"type": kind}
	for _, field := range fields {
		if value, present := obj[field]; present {
			out[field] = value
		}
	}
	return kind, out, nil
}

// Canonicalize returns the RFC 8785 identity-bearing bytes of an object.
func Canonicalize(obj map[string]any, kind string) ([]byte, error) {
	_, bearing, err := IdentityBearing(obj, kind)
	if err != nil {
		return nil, err
	}
	canonical, err := SerializeJCS(bearing)
	if err != nil {
		return nil, err
	}
	return []byte(canonical), nil
}

// Identify returns the content-addressed identifier:
// scheme + ":" + SHA-256 hex of the canonical identity-bearing bytes.
func Identify(obj map[string]any, kind string) (string, error) {
	resolvedKind, bearing, err := IdentityBearing(obj, kind)
	if err != nil {
		return "", err
	}
	canonical, err := SerializeJCS(bearing)
	if err != nil {
		return "", err
	}
	digest := sha256.Sum256([]byte(canonical))
	return Prefix[resolvedKind] + ":" + hex.EncodeToString(digest[:]), nil
}

File: bindings/csharp/Causalontology/Canonical.cs
// Canonicalization and content-addressed identity (spec/identity.md):
//   1. take the object as JSON,
//   2. keep only the identity-bearing fields for its kind (with "type" injected),
//   3. serialize with the JSON Canonicalization Scheme (RFC 8785),
//   4. hash with SHA-256,
//   5. identifier = scheme + ":" + lowercase hex digest.

using System.Security.Cryptography;
using System.Text;

namespace Causalontology;

public static class Canonical
{
    // The identity-bearing fields of each of the seventeen kinds. "type" is
    // always injected, so it is not listed here. Order does not matter (JCS
    // sorts keys). 2.0.0: whole-word re-mint (Principle P7).
    public static readonly IReadOnlyDictionary<string, string[]> IdentityFields =
        new Dictionary<string, string[]>
        {
            // ---- type tier ----
            ["occurrent"] = new[] { "label", "category", "stratum" },
            ["causal_relation_object"] = new[] { "causes", "effects",
                "mechanism", "temporal", "modality", "context", "refines",
                "skips" },
            ["continuant"] = new[] { "label", "category" },
            ["realizable"] = new[] { "kind", "bearer", "label" },
            ["stratum"] = new[] { "label", "scheme", "ordinal", "unit",
                "governs" },
            ["bridge"] = new[] { "coarse", "fine", "relation" },
            ["port"] = new[] { "bearer", "label", "direction", "accepts",
                "realizable" },
            ["conduit"] = new[] { "label", "from", "to", "carries",
                "transform" },
            ["quality"] = new[] { "label", "datatype", "unit", "stratum" },
            // ---- token tier ----
            ["token_individual"] = new[] { "instantiates", "designator",
                "part_of" },
            ["token_occurrence"] = new[] { "instantiates", "interval",
                "participants", "locus", "observer" },
            ["state_assertion"] = new[] { "subject", "quality", "value",
                "interval" },
            ["token_causal_claim"] = new[] { "causes", "effects",
                "covering_law", "actual_delay", "counterfactual" },
            // ---- provenance tier ----
            ["assertion"] = new[] { "about", "source", "evidence_type",
                "evidence", "strength", "confidence", "timestamp",
                "evidenced_by" },
            ["enrichment"] = new[] { "about", "field", "entry", "source",
                "timestamp" },
            ["retraction"] = new[] { "retracts", "source", "timestamp" },
            ["succession"] = new[] { "predecessor", "successor", "timestamp" },
        };

    // Whole-word re-mint (P7): the scheme IS the type value for every kind.
    public static readonly IReadOnlyDictionary<string, string> Prefix =
        IdentityFields.Keys.ToDictionary(k => k, k => k);

    public static readonly IReadOnlyDictionary<string, string> KindOfPrefix =
        Prefix.ToDictionary(kv => kv.Value, kv => kv.Key);

    /// <summary>Infer an object's kind from its type field, id prefix, or shape.</summary>
    public static string InferKind(JsonMap obj)
    {
        if (obj.Get("type") is string type)
            return type;
        if (obj.Get("id") is string id && id.Contains(':'))
        {
            var pre = id.Split(':', 2)[0];
            if (KindOfPrefix.TryGetValue(pre, out var kind))
                return kind;
        }
        if (obj.ContainsKey("coarse") && obj.ContainsKey("fine"))
            return "bridge";
        if (obj.ContainsKey("causes") && obj.ContainsKey("effects"))
            return "causal_relation_object";
        if (obj.ContainsKey("retracts"))
            return "retraction";
        if (obj.ContainsKey("predecessor") && obj.ContainsKey("successor"))
            return "succession";
        if (obj.ContainsKey("field") && obj.ContainsKey("entry"))
            return "enrichment";
        if (obj.ContainsKey("evidence_type")
            || (obj.ContainsKey("about") && obj.ContainsKey("confidence")))
            return "assertion";
        if (obj.ContainsKey("kind") && obj.ContainsKey("bearer"))
            return "realizable";
        throw new ArgumentException(
            "cannot infer kind (occurrents and continuants share a shape); "
            + "pass kind explicitly");
    }

    /// <summary>The identity-bearing subset of an object, with type always present.</summary>
    public static (string Kind, JsonMap Subset) IdentityBearing(
        JsonMap obj, string? kind = null)
    {
        kind ??= InferKind(obj);
        if (!IdentityFields.TryGetValue(kind, out var fields))
            throw new ArgumentException($"unknown kind: {kind}");
        var subset = new JsonMap { { "type", kind } };
        foreach (var field in fields)
        {
            if (obj.ContainsKey(field))
                subset[field] = obj[field];
        }
        return (kind, subset);
    }

    /// <summary>The RFC 8785 identity-bearing bytes of an object.</summary>
    public static byte[] Canonicalize(JsonMap obj, string? kind = null)
    {
        var (_, subset) = IdentityBearing(obj, kind);
        return Encoding.UTF8.GetBytes(Jcs.Serialize(subset));
    }

    /// <summary>The content-addressed identifier: scheme + ':' + SHA-256 hex.</summary>
    public static string Identify(JsonMap obj, string? kind = null)
    {
        var (resolvedKind, subset) = IdentityBearing(obj, kind);
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(Jcs.Serialize(subset)));
        return Prefix[resolvedKind] + ":" + Convert.ToHexString(digest).ToLowerInvariant();
    }
}
